### ISSUE TITLE
Proposal for support in remote segment deletion

### ISSUE THREAD

**Issue by jamesyfshao** [linkedin/pinot#2248]
## Motivation
In some remote segment push job for offline tables, user store offline segments in deep storage (HDFS, S3, etc) and let pinot server to download them from deep storage with the appropriate URI address. However, pinot controller doesn't automatically do any clean up for the segment stored in deep storage while it does clean up for expired segment metadata on zookeeper. In the long run having large amount of unused segment files on deep storage could cause storage issue and incur higher hardware cost. One better way to handle this issue is to allow user to implement segment deletion behaviors in their segment fetchers and execute the segment deletion for deep storage periodically while we do clean up for segment zk meta data

## Proposed changes
To support for this features, we will need to add a method to `SegmentFetcher` interface. We will add a method `void removeSegmentFromRemote(String uri) throws IOException;` to the current list of methods of this interface:

	public interface SegmentFetcher {
	  void init(Configuration configs);
	  void fetchSegmentToLocal(String uri, File tempFile) throws Exception;
	  void removeSegmentFromRemote(String uri) throws IOException;
	}

`removeSegmentFromRemote` method will delete the segment for the given uri upon call, and throws IOException if such deletion failed. For segment fetcher that doesn't support deletion, they can just implement an empty method.

In pinot controller, we will add logics to invoke the this deletion method in `SegmentDeletionManager`. In method `deleteSegmentFromPropertyStoreAndLocal` we will try to fetch the segment node in property store and then delete the segment in deep storage with its uri information stored on zk. We will lookup the proper segmentFetcher for each segment uri and invoke the proper deletion method for it.

If the deletion failed (by throwing IOException), we will log the error within `SegmentDeletionManager` and continue with other deletion operations. Users can set up proper monitor for deep storage deletion failure and perform appropriate processing for it.

**Comment by kishoreg** [comment]
this is a good proposal. quick feedback, should we introduce the concept of FileSystem instead of having SegmentFetcher?

**Comment by jamesyfshao** [comment]
@kishoreg I definitely think that would make more sense. However, this might break certain users' segment fetcher as the current pinot logics allow users to implement their own extension of segment fetcher.

**Comment by jamesyfshao** [comment]
@jfim I agree this would potentially break users' expectation of Pinot behavior on deep storage. However, storing data on S3/HDFS without automatic clean up option could translate to higher storage cost for use cases that don't care much about pinot deep storage, especially given most of these files are just pinot segment files and usually not the source of truth at the moment. 

I would suggest that we should provide a configuration option to users to decide whether to perform deletion on deep storage (with the default being False and detail warning documentation). This feature would also require implementation of deletion logic in each customized segmentFetcher, so users could provide their own logic to handle proper deletion logics/configuration if they are loading from a customized deep storage location.

**Comment by jamesyfshao** [comment]
@jfim @kishoreg any comments on this proposal? Would you guys want to see a draft implementation and we can chat about whether it makes sense?

**Comment by kishoreg** [comment]
@jamesyfshao Yes, a draft on how to support local file system and the remote file system will definitely help. Please add some information on how the code should be structured. For e.g If we want to support for HDFS, S3, Azure -  where do we put the code and what are the configuration parameters?
